### DISCUSSION TITLE
Need to obtain shadow free images

### DISCUSSION THREAD

**Discussion by TolgaAktas** [torchgeo/torchgeo#1160] [Q&A]
I am looking for a way to get dataloader to fetch a triplet of cloud-free, clouded and cloud mask images of a given area. 

1. What do you think might be the best strategy to modify the __get_item__ function? 
2. Do you think we can also sample by the the cloud cover ratio? 
3. Does torchgeo make use of the information in MTL file?

**Comment by adamjstewart** [comment]
1. Depends on exactly what you're looking for. If you're looking for both cloud-free and clouded images of the same area, I assume you don't care about the time component? You could create one GeoDataset containing all cloud-free images and a second GeoDataset containing clouded images and cloud masks, then compute the intersection between these two datasets and sample from that. You would need to ignore the time dimension by not specifying it in the filename_regex. 
2. We don't currently store metadata about cloud cover in the R-tree, but you could, and then design a custom sampler that weights which image is chosen based on that metadata.
3. We do not currently make use of information in MTL files in any of our existing datasets, although it wouldn't be hard to add a custom dataset that does (I believe it's just an XML file).

**Reply by TolgaAktas** [reply]
Option 1 sounds smart, I totally forgot about the intersection. But I am not intending to mix up datasets, I am rather trying to serve the triplets of the matching area. Time component doesn't matter as you pointed, but I am looking for the pixel-wise matching accuracy. 

How would I go about Option 2?

**Reply by adamjstewart** [reply]
Option 1 will give you pixel-wise matching images, you don't have to be mixing datasets. For example:
```python
cloud_free = Landsat8(root="/directory/containing/cloud_free/images")
clouded = Landsat8(root="/directory/containing/clouded/images", bands=Landsat8.default_bands + ["pixel_qa"])
dataset = cloud_free & clouded
```

For option 2, look at the implementation of `RasterDataset.__init__`. Specifically, on [this line](https://github.com/microsoft/torchgeo/blob/v0.4.0/torchgeo/datasets/geo.py#L362), instead of inserting only the filename, you can insert a tuple of (filename, cloud_cover_ratio). The `RasterDataset.__getitem__` would need to be overridden since the index now contains two things instead of one.

Then, you would need to implement your own custom sampler. Check out the implementations in `torchgeo.samplers`. `RandomGeoSampler` and `RandomBatchGeoSampler` both draw samples weighted by the area of each entry in the index. You would instead weight based on cloud cover ratio.

**Reply by TolgaAktas** [reply]
I am thinking of feeding cloud_free images as the ground truth, output label however. I think the intersection would be very convenient, it if is permitting me to fetch my input and output images from the intersection. Do you think that could be an outcome for using intersection?

**Reply by adamjstewart** [reply]
Yes, IntersectionDataset is designed solely for the exact purpose you describe. The sampler is only allowed to sample from regions that appear in both datasets, and it will sample from both with the exact same bounding box, resolution, coordinate system, etc.

**Reply by TolgaAktas** [reply]
Do you know how the channels from two images from two datasets are combined? I am doing the following, but the samples from both images are identical for some reason. np.allclose return True

```python
bands = ["B1","B2","B3","B4","B5","B6","B7","B8","B9","B10", "B11","QA_PIXEL"]
dataset = Landsat8(root=root_path,bands=bands)
dataset.rgb_bands = ["B4","B3","B2"]

cloudfree_dataset = Landsat8(root=cloudfree_path,bands=bands)
cloudfree_dataset.rgb_bands = ["B4","B3","B2"]

all_dataset = dataset & cloudfree_dataset
all_dataset.rgb_bands = ["B4","B3","B2"]
all_dataset.bands = bands

for elem in dataloader:
    break

idx = 0

sample1 = img[idx,:12].float()
sample2 = img[idx,12:].float()

np.allclose(sample1,sample2)
```

**Reply by adamjstewart** [reply]
They should just be stacked along the channel dimension. Are you sure `root_path` and `cloudfree_path` don't contain any of the same images? It's not clear from your code where `img` comes from. Can you share enough code and maybe some example images that I can reproduce the problem on my system?

P.S. You shouldn't need to override `rgb_bands` or `all_bands` to use the datasets.

**Reply by TolgaAktas** [reply]
for 0.3.1, without bands definition, it doesn't recognize a valid Landsat8 dataset from the directory. They shouldn't contain any similar images at all.

**Reply by adamjstewart** [reply]
I would recommend using 0.4.0 if possible. However, 0.4.0 contains a known bug that was fixed in #1135. This will be included in the 0.4.1 release, coming soonish.
